Problem: HLC invariants can be broken (#32) and too much use of static

omarkj opened the pull request:
Solution: This PR introduces a memory mapped meta database that is used by modules to persist whatever state they might need to reconstruct their previous state after a restart. For now this is only done for HLCs (see 77b2506).

Additionally, remove excessive usage of `lazy_static` (06e1156, 5200d65, 4c7065f).

Commit:
Problem: Too many statics (`PUBLISHER`)

We use `lazy_static` as a clutch too much as it’s
not really recommended way of working in rust if
it is possible to prevent it.

Solution: Remove the `PUBLISHER` static as it is
not needed.

Commit:
Problem: Too many statics (`STORAGE`)

We use `lazy_static` as a clutch too much as it’s
not really recommended way of working in rust if
it is possible to prevent it.

Solution: Remove the `STORAGE` static as it is
not needed.

To do this, the thread needs to move a reference
to the storage into it’s context as it is spawned,
the previous design didn’t allow for this as
easily since the `sender` had to be returned to
spawning thread. To solve this a single-use
channel is used to communicated the newly created
sender to the spawning thread. This channel will
only be used for that one purpose afterwards no
references will exist pointing to it.

Commit:
Problem: Too many statics (`HLC_CLOCK`)

We use `lazy_static` as a clutch too much as it’s
not really recommended way of working in rust if
it is possible to prevent it.

Solution: Remove the `HLC_CLOCK` static as it is
not needed.

The only way that I can see having spent
considerable time reading over Rust’s
documentation on sharing references safely between
threads is by using `Arc`, or a smart pointer that
uses atomic operations to count references.

This will add some memory overhead (the amount
escapes me, but it is similar to `Rc` or two
`usize` values) as well as atomic reference
counting overhead but those are only done during
cloning and when the last `Arc` goes out of scope.
In PDB’s situation the last `Arc` will only go out
of scope once the server is shut down. Also, we do
not dynamically send the timestamp between
threads, it only happens during startup for a
constant number of threads (that constant is not
the same between instances, tho).

As before, the `HLC` clock is protected with a
`Mutex`.

Commit:
Problem: HLC guarantees can be broken by sys clock

The system clock can go backwards in multiple
cases, if it has skewed to far forward for
instance. This will break the HLC promises since
the HLC timestamp is generated using the system
clock.

Solution: Introduce a `scratchpad` which is a meta
database internal to PumpkinDB. This file is
memory mapped into memory and individual modules
have access to a subset of the memory mapped file.
For now, only the `timestamp` module uses it.

Multiple threads _can not_ write to the same piece
of memory (obviously) so writes need to guarantee
one writer. For now this is done with a Mutex.

This adds a dependency on a new crate, `memmap`.

omarkj commented:
Solves https://github.com/PumpkinDB/PumpkinDB/issues/32

omarkj commented:
I'm running the same version as TravisCI here locally but not having any issues. TravisCI's logs seem to indicate it doesn't even try to download the dependencies.. So it's not failing on the actual changes..

yrashk commented:
Actually, AppVeyor is failing the same way Travis CI does.

yrashk commented:
I can reproduce the problem locally at least partially:

```
➜  pumpkindb git:(omarkj-omarkj-statics-meta-hlc) ✗ cargo version
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:323
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

yrashk commented:
Actually, the above issue is also present on master.

yrashk commented:
So it looks like nightly-2017-03-12 broke this, last version before that (nightly-2017-03-04) works fine.

yrashk commented:
Confirmed on `#rust` IRC, a bug.

yrashk commented:
@omarkj try rebasing off latest master. Hopefully should fix the issue.

yrashk commented on src/timestamp.rs:
```diff
@@ -6,26 +6,50 @@
+        let mut clock = self.clock.lock().unwrap();
+        let now = clock.0.now();
+        let ref mut state = clock.1;
+        let _ = unsafe { &now.write_bytes(&mut state.as_mut_slice()).unwrap() };
```
How often would the mmap be flushed?

omarkj replied:
Right now there is no guarantee when it's flushed. This is unacceptable for a production system. We should look into using [`flush_async`](https://osnacs.github.io/plantex/memmap/struct.MmapViewSync.html#method.flush_async) but I'd like that to be done later, preferably with a benchmark on the HLC.

yrashk replied:
Lets do it later, then — but lets create a ticket for this so we don't lose it.

yrashk commented:
Another issue:

```
$ RUST_BACKTRACE=1 cargo run --bin pumpkindb
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/pumpkindb`
thread 'main' panicked at 'Could not open or create scratchpad: Error { repr: Os { code: 2, message: "No such file or directory" } }', src/libcore/result.rs:860
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: <core::result::Result<T, E>>::expect
   1: pumpkindb::prepare_mmap
   2: pumpkindb::main
```

yrashk commented on src/main.rs:
```diff
@@ -21,39 +21,50 @@ pub mod pubsub;
+    scratchpad_pathbuf.push(filename);
+    scratchpad_pathbuf.set_extension("dat");
+    let scratchpad_path = scratchpad_pathbuf.as_path();
+    let scratchpad_file = OpenOptions::new().create_new(false).write(true)
```
Shouldn't this be `true` here for `create_new`?

yrashk replied:
Actually, `true` isn't right either. It would work for the new one, but then it is resulting `File exists`. Therefore, it should be `create(true)`, I think.

yrashk commented:
Also, I think we should document the mmap'ed file layout somewhere.

yrashk commented:
Also, it looks like the `sender_sender` part of the transition is quite complicated. As per our chat on Gitter, it is perhaps best to explore other ideas, like creating a receiver outside of the Scheduler and passing it over to Scheduler.

Commit:
Problem: PDB crashes if scratchpad doesn’t exist

I was using `OpenOptions` wrong here, and the
system refused to start up if the scratchpad did
not exist.

Solution: Change `create_new` to `create`, which
will not fail if the file already exists.

Commit:
Problem: Too much complexity in getting sender out

The use of a temporary channel is useless since
reciever implements `Send`able.

Solution: Instead of creating a temporary channel
to send the Sender back to pid 0, create the
sender/receiver pair before and pass the receiver
to the scheduler, retaining the sender.

Commit:
Problem: The structure of the scratchpad is hidden

Solution: Document what bytes are reserved for
what usage in code.

omarkj commented:
Updated: Some issues passing the tests on Windows. I do not have a Windows license so cannot check it, but @yrashk is looking into it.

Commit:
Problem: Windows clock is failing HLC guarantees

The Windows clock has a low resolution (100ns)
causing the HLC guarantees to fail during the
doctests. This was discovered after the removal of
static since a new clock was generated between
each test in the new implementation versus a
global clock before.

Solution: This is a temporary patch pending fixes
in upstream libraries to the Windows clock
resolution. For now a global clock is created for
the doctests and it is passed to the eval
function. This is not ideal.

omarkj commented:
After much debugging with @yrashk we found the issue is in Windows clock resolution. See #184
